test/e2e/prometheus/prometheus.go

```diff
@@ -0,0 +1,185 @@
+package prometheus
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/celestiaorg/knuu/pkg/instance"
+	"github.com/celestiaorg/knuu/pkg/knuu"
+)
+
+const (
+	instanceName          = "prometheus"
+	image                 = "prom/prometheus:v2.49.0"
+	apiPort               = 9090
+	defaultScrapeInterval = 15 * time.Second
+	prometheusConfig      = "/etc/prometheus/prometheus.yml"
+	prometheusArgs        = "--config.file=" + prometheusConfig
+)
+
+type ScrapeJob struct {
+	Name     string
+	Target   string
+	Interval time.Duration
+}
+
+type Prometheus struct {
+	knuu                *knuu.Knuu
+	instance            *instance.Instance
+	prometheusProxyHost string
+	scrapeJobs          []ScrapeJob
+}
+
+type MetricFilter struct {
+	MetricName string
+	Labels     map[string]string
+}
+
+func New(kn *knuu.Knuu) (*Prometheus, error) {
+	ins, err := kn.NewInstance(instanceName)
+	if err != nil {
+		return nil, err
+	}
+	return &Prometheus{knuu: kn, instance: ins}, nil
+}
+
+func (p *Prometheus) AddScrapeJob(ctx context.Context, scrapeJob ScrapeJob) error {
+	if p.instance.IsState(instance.StateStarted) {
+		return ErrPrometheusAlreadyRunning
+	}
+
+	if scrapeJob.Name == "" {
+		return ErrEmptyScrapeJobName
+	}
+	if scrapeJob.Target == "" {
+		return ErrEmptyScrapeJobTarget
+	}
+	p.scrapeJobs = append(p.scrapeJobs, scrapeJob)
+	return nil
+}
+
+func (p *Prometheus) Destroy(ctx context.Context) error {
+	return p.instance.Execution().Destroy(ctx)
+}
+
+func (p *Prometheus) Start(ctx context.Context) error {
+	if err := p.instance.Build().SetImage(ctx, image); err != nil {
+		return err
+	}
+
+	if err := p.instance.Network().AddPortTCP(apiPort); err != nil {
+		return err
+	}
+
+	if err := p.instance.Build().Commit(ctx); err != nil {
+		return err
+	}
+
+	if err := p.instance.Storage().AddFileBytes([]byte(p.generateConfig()), prometheusConfig, "0:0"); err != nil {
+		return err
+	}
+
+	if err := p.instance.Build().SetArgs(prometheusArgs); err != nil {
+		return err
+	}
+
+	if err := p.instance.Execution().Start(ctx); err != nil {
+		return err
+	}
+
+	proxyHost, err := p.instance.Network().AddHost(ctx, apiPort)
+	if err != nil {
+		return err
+	}
+	p.prometheusProxyHost = proxyHost
+
+	return nil
+}
+
+// GetMetric returns the first matching metric from the prometheus instance
+// Example:
+// GetMetric(MetricFilter{MetricName: "node_start_ts", Labels: map[string]string{"instance_name": "12D3KooWNHuCns..."}})
+func (p *Prometheus) GetMetric(filter MetricFilter) (float64, error) {
+
+	labels := []string{}
+	for k, v := range filter.Labels {
+		labels = append(labels, fmt.Sprintf(`%s="%s"`, k, v))
+	}
+	query := fmt.Sprintf("%s{%s}", filter.MetricName, strings.Join(labels, ","))
+
+	resp, err := http.Get(fmt.Sprintf("%s/api/v1/query?query=%s", p.prometheusProxyHost, query))
```

Comment:
`http.Get` doesn't have a timeout. So, the operation can block forever.

Reply:
using a function that has a timeout now